Source/Interceptor.cs

```diff
@@ -170,41 +170,33 @@ public override bool Equals(object obj)
 
 				var eq = key.fixedString == this.fixedString && key.values.Count == this.values.Count;
 
-                //the code below is broken as it uses an OR when checking the arguments, this means that if any pair of arguments match 
-                //the result is a match.
-                //This is only going to hit some edge cases as for the most part the fixed string above spots arguments correctly.
-                //Fixing this really needs a reworking of the GetHashCode, and also some sorting out of how to compare value types that have been
-                //boxed correctly (another problem this code has)
-                var index = 0;
-                while (eq && index < this.values.Count)
-                {
-                    eq |= this.values[index] == key.values[index];
-                    index++;
-                }
-
-                return eq;
+				var index = 0;
+				while (eq && index < this.values.Count)
+				{
+					// using `object.Equals` instead of == ensures that we get the correct
+					// comparison result for boxed value types:
+					eq &= object.Equals(this.values[index], key.values[index]);
+					index++;
+				}
+
+				return eq;
 			}
 
-            public override int GetHashCode()
-            {
-                var hash = fixedString.GetHashCode();
-
-                var factor = 1;
-                foreach (var value in values)
-                {
-                    if (value != null)
-                    {
-                        // we use a factor that increases with each following value (argument)
-                        // so that if the values are in a different order, we get a different hash code
-                        // see GitHub issue #252
-                        hash ^= value.GetHashCode() / factor;
-                    }
-                    factor *= 3;
-                }
-
-                return hash;
-            }
-        }
+			public override int GetHashCode()
+			{
+				var hash = fixedString.GetHashCode();
+
+				foreach (var value in values)
+				{
+					if (value != null)
+					{
+						hash = unchecked((hash * 397) ^ value.GetHashCode());
```

Comment:
We don't care whether the hash value overflows or not, but we do not want overflows to trigger an exception. That is why `unchecked` is necessary here.